wasm/src/validator.rs

```diff
@@ -0,0 +1,400 @@
+//! Contains functional related to validators
+
+use alloc::string::String;
+
+use crate::data_model::{permission::validator::Verdict, prelude::*};
+
+pub mod prelude {
+    //! Contains useful re-exports
+
+    pub use iroha_data_model::{permission::validator::Verdict, prelude::*};
+    pub use iroha_wasm_derive::{validator_entrypoint as entrypoint, Token, Validate};
+
+    pub use super::traits::{Token, Validate};
+    #[cfg(feature = "debug")]
+    pub use crate::DebugExpectExt as _;
+    pub use crate::{deny, pass, pass_if, validate_grant_revoke, EvaluateOnHost as _};
+}
+
+pub mod macros {
+    //! Contains useful macros
+
+    /// Shortcut for `return Verdict::Pass`.
+    #[macro_export]
+    macro_rules! pass {
+        () => {
+            return $crate::data_model::permission::validator::Verdict::Pass
+        };
+    }
+
+    /// Macro to return [`Verdict::Pass`](crate::data_model::permission::validator::Verdict::Pass)
+    /// if the expression is `true`.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// pass_if!(asset_id.account_id == authority);
+    /// ```
+    #[macro_export]
+    macro_rules! pass_if {
+        ($r:expr) => {
+            if $r {
+                return $crate::data_model::permission::validator::Verdict::Pass;
+            }
+        };
+    }
+
+    /// Shortcut for `return Verdict::Deny(...)`.
+    ///
+    /// Supports [`format!`](alloc::format) syntax as well as any expression returning [`String`](alloc::string::String).
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// deny!("Some reason");
+    /// deny!("Reason: {}", reason);
+    /// deny!("Reason: {reason}");
+    /// deny!(get_reason());
+    /// ```
+    #[macro_export]
+    macro_rules! deny {
+        ($l:literal $(,)?) => {
+            return $crate::data_model::permission::validator::Verdict::Deny(
+                ::alloc::fmt::format(::core::format_args!($l))
+            )
+        };
+        ($e:expr $(,)?) =>{
+            return $crate::data_model::permission::validator::Verdict::Deny($e)
+        };
+        ($fmt:expr, $($arg:tt)*) => {
+            return $crate::data_model::permission::validator::Verdict::Deny(
+                ::alloc::format!($fmt, $($arg)*)
+            )
+        };
+    }
+
+    /// Macro to parse literal as a type. Panics if failed.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use iroha_wasm::parse;
+    /// use iroha_wasm::data_model::prelude::*;
+    ///
+    /// let account_id = parse!("alice@wonderland" as <Account as Identifiable>::Id);
+    /// ```
+    #[macro_export]
+    macro_rules! parse {
+        ($l:literal as _) => {
```

Comment:
I think `_` is a pattern so it wouldn't be matched as a type. My suspicion is this is useless, please confirm

Reply:
No, this really works, I tryed. Also take a look at this [syn docs](https://docs.rs/syn/latest/syn/enum.Type.html#variant.Infer). Not a proof, but makes sense